### PR TITLE
Bintools Wrapper: Init by refactoring out of cc-wrapper, take 2

### PR DESCRIPTION
**Pull request by Ericson2314**
**edit** see my question in https://github.com/NixOS/nixpkgs/pull/29396#issuecomment-333688023

###### Motivation for this change

This is meant to be reviewed commit by commit.

This is needed so packages that just use binutils get the same setup-hook support. That is, variables like `BUILD_AS` or `LD` should be defined the same way, along with the `LD_FLAGS`.  This will be necessary to clean up at least the gcc derivation (which must be built with an binutils not associated with any existsing C compiler when target != host), and probably other compilers too.

More broadly, having two wrappers will help enforce proper layering between binutils and the C compiler. For languages besides C/C++, it would be neat to use `stdenvNoCC + binutils, avoiding a C compiler dependency. This wrapper makes that change possible.

@edolstra I'm hoping for you to merge this change when you approve of it, to avoid a repeat of last time.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built stdenv on platform(s)
   - [ ] NixOS --- in progress
   - [x] macOS
   - [ ] Linux 
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/default.nix
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: @edolstra this symlink avoids the incompatibility the old version introduced.

File: pkgs/build-support/cc-wrapper/ld-wrapper.sh
Comment (orivej) [inline pkgs/build-support/cc-wrapper/ld-wrapper.sh]: It was determined to be valuable not to iterate over params more times than necessary. Could you merge this loop into the next one? (It has already merged two cases, `[ "$NIX_@infixSalt@_DONT_SET_RPATH" != 1 ]` and `[ "$NIX_@infixSalt@_SET_BUILD_ID" = 1 ]`.)

Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/ld-wrapper.sh]: Good idea!

Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/ld-wrapper.sh]: Done

File: pkgs/build-support/binutils-wrapper/ld-wrapper.sh
Comment (orivej) [inline pkgs/build-support/binutils-wrapper/ld-wrapper.sh]: These cases are not related, so if you enumerate them, they should be 1 and 2.

File: pkgs/build-support/binutils-wrapper/setup-hook.sh
Comment (orivej) [inline pkgs/build-support/binutils-wrapper/setup-hook.sh]: Where may `depOffset` be set to 1? I find only places where it is set to -1 or 0.

File: pkgs/build-support/binutils-wrapper/ld-wrapper.sh
Comment (Ericson2314) [reply pkgs/build-support/binutils-wrapper/ld-wrapper.sh]: True. Good catch.

File: pkgs/build-support/binutils-wrapper/setup-hook.sh
Comment (Ericson2314) [reply pkgs/build-support/binutils-wrapper/setup-hook.sh]: Oh heh. That would anticipate https://github.com/NixOS/nixpkgs/pull/26805

File: pkgs/build-support/binutils-wrapper/ld-wrapper.sh
Comment (Ericson2314) [reply pkgs/build-support/binutils-wrapper/ld-wrapper.sh]: Done

File: doc/stdenv.xml
Comment (DavidEGrayson) [inline doc/stdenv.xml]: "Sine" is a typo.

Comment (Ericson2314) [reply doc/stdenv.xml]: Thanks!

File: pkgs/build-support/binutils-wrapper/setup-hook.sh
Comment (Ericson2314) [reply pkgs/build-support/binutils-wrapper/setup-hook.sh]: Cc-wrapper already has that, I guess I'll leave it as consistent anticipation.

File: doc/stdenv.xml
Comment (DavidEGrayson) [inline doc/stdenv.xml]: Probably you're missing the words "be added to" in this sentence.

File: doc/stdenv.xml
Comment (DavidEGrayson) [inline doc/stdenv.xml]: The idomatic wording options would be:

- "may be in use simultaneously", or
- "may be in simultaneous use"

File: doc/stdenv.xml
Comment (DavidEGrayson) [inline doc/stdenv.xml]: "the" should be "that"

File: doc/stdenv.xml
Comment (DavidEGrayson) [inline doc/stdenv.xml]: "actually" is not an adjective, probably you meant "actual".

**Comment by Ericson2314** [comment]
[Blocked by https://github.com/NixOS/nixpkgs/pull/29568]

**Comment by Ericson2314** [comment]
I'm rebuilding now, but I've built other PRs containing this so quite confident it will succeed. This is ready for final review!

File: pkgs/build-support/binutils-wrapper/setup-hook.sh
Comment (Ericson2314) [inline pkgs/build-support/binutils-wrapper/setup-hook.sh]: @edolstra added a commit to fix the `CMD=windres` issue you pointed out in https://github.com/NixOS/nixpkgs/issues/29846 .

**Comment by Ericson2314** [comment]
On second thought, I'd like to rename this `utils-wrapper`, and `stdenv.cc.utils`. "cctools + GNU binutils" is also wrapped, and isn't "GNU Binutils" so I think a name like "utils" would be better.

**Comment by copumpkin** [comment]
@Ericson2314 that seems fine to me

**Comment by Ericson2314** [comment]
I'm going to wait to kick of a nixborg mass rebuild until https://github.com/NixOS/nixpkgs/pull/31775 is finished, but this is now again ready for rebuild.

**Comment by Ericson2314** [comment]
@nixborg buidl

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

File: pkgs/build-support/cc-wrapper/add-flags.sh
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/add-flags.sh]: I tried to make this use `CFLAGS_LINK` instead of `LDFLAGS` for better layering, but that changed the priority breaking stdenv (as the bootstrap tools lib directory will resolve so much).

**Comment by Ericson2314** [comment]
I built Atlas locally (without any rebase, unlike nixborg) and it succeeded. This reassures me the build failure is spurious / from something else, not a problem with the wrapper and `gfortran` as I feared.

**Comment by Ericson2314** [comment]
I've add a few more cleanups on top, doing another rebuild.

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by Ericson2314** [comment]
Curious whether the last `LDEMULATION`-defining commit unbroke cross, so repurposed https://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-trunk . If not, I'll just give up on `--enable-targets=all` in binutils: it's less a panacea than I thought because the GNU Assembler doesn't become multi-target.

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

File: pkgs/os-specific/darwin/binutils/default.nix
Comment (Ericson2314) [inline pkgs/os-specific/darwin/binutils/default.nix]: @vcunat there's no nix-level way to do per-output propgated deps, right?

**Comment by orivej** [comment]
> If not, I'll just give up on `--enable-targets=all` in binutils

I think multitarget binutils (enabled in https://github.com/NixOS/nixpkgs/pull/30484/commits/f8741c38cd546e3ff18ce9d708de14ff2aae68ab) is a good thing even if they can not be used for cross.

File: pkgs/os-specific/darwin/binutils/default.nix
Comment (vcunat) [reply pkgs/os-specific/darwin/binutils/default.nix]: @Ericson2314: right, I think there's no better way than this ATM.  There certainly wasn't when multi-output stdenv was done (years ago), and I haven't seen any use case up to now.  If more of them appear, we can design some kind of API.

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/26805 is needed to fix progated inputs, so there's little point trying cross with this PR alone. I moved the `LDEMULATION` commit to that other PR accordingly. I also pullled  some non-controversial stuff out of the end of this into https://github.com/NixOS/nixpkgs/issues/28448 into https://github.com/NixOS/nixpkgs/issues/28448.

I'm looking to merge this within the week, so I can get onward to the more interesting cross stuff. I now count 3 separate instances of people decently blocked on me unbreaking things, and this has sat open for a while. @edolstra you're the point person for a change this big, so let me know what I can do to help you review.

**Review by vcunat — APPROVED** [review]
OK, I think.

I read though this all, not finding any problem (beyond the nitpicks I commented on); but I can't say I fully understand _all_ the necessary bits.

**Comment by orivej** [comment]
These breaks the build of `libstdcxx5`.

File: pkgs/build-support/cc-wrapper/utils.sh
Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/utils.sh]: That's an oversight, will fix.

File: pkgs/build-support/bintools-wrapper/default.nix
Comment (Ericson2314) [reply pkgs/build-support/bintools-wrapper/default.nix]: What do you mean by `programs.man.enable`? I'll gladly just remove `outputsToInstall`, or do whatever you think is best, just confused what that is.

File: doc/stdenv.xml
Comment (Ericson2314) [reply doc/stdenv.xml]: Thanks!

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@grahamc what is the `nix-instantiate ...` that's failing? the ellipses won't expand.

**Comment by grahamc** [comment]
If you triple-click the text it'll select it all and you can copy it out: `grahamcofborg-eval-nixpkgs-unstable-jobset — nix-instantiate ./pkgs/top-level/release.nix -A unstable `

**Comment by Ericson2314** [comment]
@grahamc thanks!

**Comment by Ericson2314** [comment]
Ah, I did `dynamic-linker-32` not `dynamic-linker-m32` in the fix. @orivej I think I fixed those issues you mentioned.

**Comment by Ericson2314** [comment]
@nixborg build

**Comment by nixborg** [comment]
Jobset created at https://hydra.mayflower.de/jobset/nixos/pr-29396

**Comment by Ericson2314** [comment]
@orivej both are now fixed

**Comment by Ericson2314** [comment]
@edolstra told me he doesn't have time to review this any time soon, so we might as well not hold it up further. Yay, thanks!

This has passed nixborg CI many times before. the last change is a mass rebuild, but just from merging with staging. The post-merge change is just an eval fix for a newer package (clang multilib). Merging, but will keep an I on staging just in case.

**Comment by orivej** [comment]
This bloke `bloaty`, fixed in fe61c3b84e8e81a8ec2bf6b3ed2a0e8652cea190.

**Comment by orivej** [comment]
This broke `lispPackages.cl-fuse`. @Ericson2314 could you fix it? It exports `NIX_LDFLAGS` with `-L/nix/store/wr579906xik1qbnblm0wq1nissxh53sy-fuse-2.9.7/lib` and runs `gcc` (wrapper), but `gcc` invokes unwrapped `gcc` without these `NIX_LDFLAGS`. Here is this sequence of steps: https://gist.github.com/orivej/4c8fc1b12d8d630edf24b6ff5f88cd80 (`141-142-sbcl` runs `sbcl`, which runs `gcc` in the same way as `142-143-gcc`, which runs the unwrapped `gcc` just like `143-144-gcc `.) Here is how `NIX_LDFLAGS` are defined: https://github.com/NixOS/nixpkgs/blob/ea15e43693908291b48f3852c32721a085aa8f84/pkgs/development/lisp-modules/define-package.nix#L23

**Comment by orivej** [comment]
Two perl packages,`perlPackages.EncodeDetect` and `perlPackages.ExtUtilsCppGuess`, fail to build with `ld: cannot find -lstdc++`. They are needed for `spamassassin` and `slic3r`. @Ericson2314 agreed to try to fix them.

It seems that `lispPackages.cl-fuse` may not be fixed before the merge to master. /cc @7c6f434c

**Comment by Ericson2314** [comment]
Can't fix this instant, but the crux of the Perl packages' problem is that `ld` no longer magically see's into the C(XX)'s compiler's lib directory. `LD=$CC` might be the best solution to avoid violating layers.

**Comment by orivej** [comment]
Thanks! I'm fixing perl packages in #32830.

**Comment by 7c6f434c** [comment]
Hm. I don't really understand where `NIX_LDFLAGS` gets lost in CL-Fuse build… So I gave up and the shared library is now prebuilt separately.

**Pull request merged**
